### DISCUSSION TITLE
Performance question about RPC

### DISCUSSION THREAD

**Discussion by CarlSayres** [activej/activej#246] [General]
I am using ActiveJ RPC to refactor an application. I will have many remote services, perhaps 40 or 50 by the time I'm done. I'm following the Key-Value example on the web site, ie.
```
public class ServerModule extends AbstractModule {
// ...
    @Provides
    RpcServer rpcServer(Eventloop eventloop) {
        return RpcServer.create(eventloop)
                .withSerializerBuilder(SerializerBuilder.create())
                .withMessageTypes(PutRequest.class, PutResponse.class, GetRequest.class, GetResponse.class)
                .withHandler(PutRequest.class, req -> Promise.of(new PutResponse(KeyValueStore.instance.put(req.getKey(), req.getValue()))))
                .withHandler(GetRequest.class, req -> Promise.of(new GetResponse(KeyValueStore.instance.get(req.getKey()))))
                .withListenPort(RPC_SERVER_PORT);
    }
}
```

The above example has just two services. What are the performance considerations if I had 50 services (or more)? Is it ok to have that many handlers multiplexed using a single RpcServer? That would mean 50 handlers and 100 message types all lumped together.
Or would there be benefits to a separate module for each service, each running on it's own port? Or grouping a few services together on a port?
I'd appreciate any guidance you can offer.
Thanks,
Carl

**Comment by eduard-vasinskyi** [comment]
Hi, @CarlSayres

There is an almost zero penalty for using multiple RPC handlers for a single RPC server. Each RPC message is effectively encoded, so multiple message types can be easily handled.

However, you may face some performance issues in case there are too many requests per second coming to the RPC server from multiple RPC clients. In that case, you could use the [Worker module](https://activej.io/boot/workers) to launch N identical RPC servers. Each server will be running inside a dedicated `Eventloop` (dedicated thread), so that the request handling load is equally distributed between worker servers.

**Comment by CarlSayres** [comment]
Hi @eduard-vasinskyi 

Thanks so much for the response.  I'm reading through the Worker module docs. That page indicates that there can be a Primary EventLoop which receives the requests which delegates to a WorkerPool. And the example given is for Http. 

Can the same work for RPC? The Memcached example in the docs takes a different approach. It create a pool of RPCServers, each with a different port #. And then it is up to the client to distribute its requests among those ports.

I would prefer a single port for RPC, and the server would distribute the requests to Worker threads. Can that work? Is there an example showing how to do this?

Thanks so much for the help!

Carl

**Reply by eduard-vasinskyi** [reply]
> Can the same work for RPC?

Yes, absolutely, All you need to do is replace worker `AsyncHttpServer` with worker `RpcServer`. 

The memcached example emulates a situation when you have multiple RPC servers on different machines. Each server listens on its own port.

If you want to have multiple servers listening on a single port you need to use Worker module and set up a primary server for distributing requests between worker servers.

I have added a simple [multithreaded RPC server launcher](https://github.com/activej/activej/blob/d98f1c03b363cabbfa291664572dc5b542ee08f3/launchers/rpc/src/main/java/io/activej/launchers/rpc/MultithreadedRpcServerLauncher.java) that you can use as a reference.

**Reply by CarlSayres** [reply]
Thank you so much!!!
